lib/service/awsoidc.go

```diff
@@ -68,13 +68,12 @@ func (process *TeleportProcess) initAWSOIDCDeployServiceUpdater() error {
 		return nil
 	}
 
-	// TODO: use the proxy channel if available?
-	// This would require to pass the proxy configuration there, but would avoid
-	// future inconsistencies: if the proxy is manually configured to serve a
-	// static version, it will not be picked up by the AWS OIDC deploy updater.
-	upgradeChannel, err := automaticupgrades.NewDefaultChannel()
-	if err != nil {
-		return trace.Wrap(err)
+	if len(channels) == 0 {
+		return trace.BadParameter("missing automatic upgrade channels")
+	}
+	upgradeChannel, ok := channels[automaticupgrades.DefaultChannelName]
+	if !ok {
+		return trace.BadParameter("missing default automatic upgrade channel")
```

Comment:
I'm afraid we might introduce a new failure mode: if the default channel is not set (either the channels slice is nil, empty, or contains other non-default channels). We also have a bunch of logic to pick the default channel in a backward-compatible way. It's supposed to run when parsing the config, however if we don't parse a fileconf (instantiate a process manually in tests) this might fail.

I think it would be safer to
```
	if channels == nil {
		channels = make(automaticupgrades.Channels)
	}
	if err = channels.CheckAndSetDefaults(); err != nil {
		return trace.Wrap(err, "validating the automatic upgrades configuration")
	}
	upgradeChannel, ok := channels[automaticupgrades.DefaultChannelName]
	if !ok {
		// this should never happen
		return trace.BadParameter("missing default automatic upgrade channel")
	}
```

Reply:
Okay, I like this. I don't think this would ever be hit outside of tests, but at least it does something that makes sense when there is no proxy upgrade channels in config for whatever reason.

Reply:
although I feel like if the channels are already non-nil it will not be thread safe to call CheckAndSet defaults on it

Reply:
alternative: just fallback to "NewDefaultChannel". In fact, why not just put this on the channels types itself:

```

// DefaultChannel returns the default upgrade channel.
func (c Channels) DefaultChannel() (*Channel, error) {
	defaultChannel, ok := c[DefaultChannelName]
	if ok && defaultChannel != nil {
		return defaultChannel, nil
	}
	return NewDefaultChannel()
}
```

I was surprised to be reminded that reading from a nil map is perfectly safe in golang, according to its spec, only nil map writes cause a panic:
> A nil map is equivalent to an empty map except that no elements may be added.

Reply:
[diff](https://github.com/gravitational/teleport/compare/b58ad4aaf6c1a58ad39a81703a8faf77ce530ac4..f511e4d0364a9de6c7187f2843a583f99c996d4f)